Add support for inelastic instruments to Discus multiple scattering correction

DannyHindson opened the pull request:
**Description of work.**

The DiscusMultipleScatteringCorrection previously only supported elastic scattering. This PR adds support for inelastic scattering following the approach in the original Discus code (which was designed for inelastic instruments). It includes support for both direct and indirect geometry instruments

The main adjustments to the code are:
- the structure factor is now a 2D function S(Q,w) instead of a 1D function S(Q). This has involved swapping the data type of the S(Q) variable from a histogram to a workspace in various places
- the logic to manage the new track properties after each scatter now generates a new k value as well as a new angle
- the elastic version of the code now expects an input workspace with units of momentum (k) instead of wavelength. While making changes to the input units to support inelastic it seemed to worth making this change (which has been desirable for a while) at the same time

**To test:**

Unzip the following zip file [DiscusTestFiles31200.zip](https://github.com/mantidproject/mantid/files/8187483/DiscusTestFiles31200.zip)
Update the path variable in the script below to point to where you unzipped the files
Run the script which carries out a calculation on some IRIS runs with a water sample (best to do on a Release build so it doesn't take too long)
Check it completes and generate a 3D surface plot for the Scatter_1 and Scatter_2 workspaces and check they contain data:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

path = 'c:\pathtofiles'

input_workspace = Load(os.path.join(path,'iris26176_graphite002_red.nxs'))
sqw = Load(os.path.join(path,'iris26176_graphite002_sqw.nxs'))

SetSample(InputWorkspace=input_workspace,
          Geometry={'Shape': 'Cylinder', 'Height': 4.0,'Radius': 1.0, 'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': 'H2-O', 'MassDensity': 1.0})

# uncomment these lines to force the calculation to be run in Direct mode. Code checks sample logs first for emode    
#input_workspace.run().addProperty("deltaE-mode", "Direct", True);
#input_workspace.run().addProperty("Ei", 1.845, True);
                                  
DiscusMultipleScatteringCorrection(InputWorkspace=input_workspace, StructureFactorWorkspace=sqw,
                                   OutputWorkspace="MuscatResults", NeutronPathsSingle=1000,
                                   NeutronPathsMultiple=1000)
```

Fixes #31200.

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Add Importance Sampling feature

Update q_dir with importance sampling logic based on inverse of a
cumulative probabilty distribution of Q.S(Q)

Remove the gaussian interpolation - was difficult to use in conjunction
with importance sampling and it also exhibited a few odd features eg
interpolation on a perfect Gaussian peak gave interpolated that weren't
same on either side of peak. Replaced with more simple "flat" interpolation

Do integration natively in the DiscusMultipleScatteringCorrection algorithm
rather than calling Integration as a child because need to get out integral
for all values on x axis so was expensive to call child algorithm.

DannyHindson committed:
Initial work on supporting inelastic mode

DannyHindson committed:
Further work on supporting inelastic mode

Extend importance sampling to inelastic by integrating S(Q, w) over
kinematically accessible region of Q, w space
Will eventually need to construct the cumulative prob distribution
per track segment so unclear at the moment whether this will be too
slow

DannyHindson committed:
Adjust look up of cumulative probability during importance sampling

Remove 3rd pseudo Q axis and just link the cumulative probability to
Q and w

DannyHindson committed:
More work on support for inelastic instruments

DannyHindson committed:
Further work on inelastic support

Implement kinc and kout based on EFixed
Retire the EMode algorithm parameter and read from workspace instead
Add a couple of unit tests for the structure factor
Add the WorkspaceCreationHelper.h to CMakeLists INC_FILES so shows up
in Visual Studio

DannyHindson committed:
Further work on inelastic support

Add switch to allow the inelastic results for different w bins
to be generated in two ways:
- reuse all track segments apart from last one
- resimulate full path for each different w bin

DannyHindson committed:
Tidy up code so elastic and inelastic behaviour aligns better

For elastic, convert the spectrum axis on S(Q) to be a DeltaE axis
with w=0 so better code sharing with inelastic
Convert the wavelength input workspace to k to make set up of required
kin and kout values more simple
Pull the code to write the results into the output workspaces back
outside of simulatePaths

DannyHindson committed:
Fix problem with InvPOfQ not being refreshed in correct place

Also convert spectrum axis to points if it's supplied as bin edges

DannyHindson committed:
Rework the inner loop to take list of energy transfers

Previously took list of kout values. As part of this change the input units
on inelastic workspaces to be energy transfer rather than wavelength
Reinstate interpolateGaussian because some inelastic workspaces have small
number of Q values (eg Discus example has only 9 Q values). Use this
where not using importance sampling

DannyHindson committed:
Small tidy ups following rebase onto main

The fact that InvPOfQ is now a workspace rather than a histogram
clashed a bit with some of the optimisations from an earlier PR
You can't have a histogram with a null E vector if the histogram
is in a workspace

DannyHindson committed:
Performance improvements

In particular calls to MatrixWorkspace.histogram(i) have been replaced
with MatrixWorkspace.getSpectrum(i) because getSpectrum returns a
reference

DannyHindson committed:
Remove missed merge conflict marker

DannyHindson committed:
Fill out inelastic documentation

Also fix bug on conversion of S(Q,w) energy transfer axis to points

DannyHindson committed:
Flip energy transfer convention to match Mantid

Update usage example to reflect this
Also add an elastic usage example
Also fiddle with conversion code to go between k and E to reduce
rounding error problems
Fix clang format problem

DannyHindson committed:
Add release note

Also fix cppcheck issues and compiler warnings
Also decide against "using namespace" in header

DannyHindson committed:
Change input units for elastic to be k rather than wavelength

More natural unit since processing is largely around determining new
output k from each scatter based on S(Q). Easier to come up with clean
examples
Also pull some code out of exec into separate functions to make exec
smaller and easier to read

DannyHindson committed:
Fix compiler warnings and add missing image

DannyHindson committed:
Adjust usage examples and fix MS concepts page

Clear workspaces at end of usage examples. RelectometryBackgroundSubtraction
crashes without this. I guess one of the leftover workspaces from the Discus
usage examples gets used as a default for the input workspace on
RelectometryBackgroundSubtraction and it doesn't like it
Fix concepts page call to DiscusMultipleScatteringCorrection
Fix compiler warnings

DannyHindson committed:
Couple of changes

Rename the NumberOfWavelengthPoints parameter that controls the interpolation
Input bins are now either Momentum or DeltaE so old name didn't make sense
Also update the indexOfValueFromCenters helper function so it can operate
without throwing an exception. This was slowing the code down by ~5 times
if an inelastic workspace with w values outside of S(Q,w) was used

DannyHindson committed:
Various changes

Update theory section of documentation to describe inelastic calculation
Functional change to update the scattering cross section and the absorption
cross section on each scatter
Add screenshot of analytical double scatter result from Mayers paper
Downgrade the messages about calls to interceptSurface and number of times
the track misses the shape to Information so not as intrusive

DannyHindson committed:
Fix two bugs

Missed a couple of formulae relating to the final track segment when
I reversed the energy transfer sign\convention
The X array of the second histogram of InvPOfQ wasn't being set -
had previously assumed setting the X array on 1st histogram was
sufficient

DannyHindson committed:
Update flag for indicating "calculate all w results"

The special value of -1.0 was previously used to indicate that results
for all w bins should be calculated from a single set of tracks in the inner
calculation loop
This doesn't work very nicely for w because this could be a real value (it
previously worked OK when the outputs required were expressed in kout values)
So update code to control results based on w index rather and value

DannyHindson committed:
Add inelastic unit tests and fix a couple of bugs
Reinstate interpolateGaussian test

DannyHindson committed:
Add more breaking changes to the release note

DannyHindson committed:
Add unit test for DeltaE interpolation

Also fix problem with interpolation step size where some bins excluded
because Einitial is negative

DannyHindson committed:
Fix small error in q integration limits

DannyHindson committed:
Update the signatures of SampleKW and GetKf

To make it easier to see what the S(Q) profile is being used for
ie w values only. Also use log S(Q) more consistently for no
importance sampling calculation

DannyHindson committed:
Correct number of default parameters in python exposed function

RichardWaiteSTFC commented on Framework/Algorithms/inc/MantidAlgorithms/DiscusMultipleScatteringCorrection.h:
```diff
@@ -51,50 +51,68 @@ class MANTID_ALGORITHMS_DLL DiscusMultipleScatteringCorrection : public API::Alg
+                                    const std::vector<double> &wValues, const Kernel::V3D &detPos,
+                                    bool specialSingleScatterCalc);
+  std::tuple<bool, std::vector<double>, double> scatter(const int nScatters, Kernel::PseudoRandomNumberGenerator &rng,
+                                                        const API::MatrixWorkspace_sptr &invPOfQ, const double kinc,
```
There is still a variable called `kinc` in this function - it has been renamed to just `k` in other functions (presumably to reflect that ki or kf could be fixed?) - should this be renamed as well?

RichardWaiteSTFC commented on Framework/Algorithms/inc/MantidAlgorithms/DiscusMultipleScatteringCorrection.h:
```diff
@@ -51,50 +51,68 @@ class MANTID_ALGORITHMS_DLL DiscusMultipleScatteringCorrection : public API::Alg
   void updateWeightAndPosition(Geometry::Track &track, double &weight, const double vmfp, const double sigma_total,
                                Kernel::PseudoRandomNumberGenerator &rng);
-  void q_dir(Geometry::Track &track, const HistogramData::Histogram &invPOfQ, const double kinc,
+  void q_dir(Geometry::Track &track, const API::MatrixWorkspace_sptr &invPOfQ, double &k,
```
The variable `kinc` in this function has been renamed to just `k` (presumably to reflect that ki or kf could be fixed?) - should this be done in other functions as well?

DannyHindson replied:
I've had a look through. Some functions such as simulatePaths and scatter do genuinely take the incident k value of a neutron path before it reaches the sample so these two are fine. There are other functions though (eg new_vector, sampleKW) that get called multiple times for a single neutron path (eg at each scatter) and there's a bit of ambiguity in my variable\parameter naming. On some occasions I've also used kinc to refer to the k value before the scatter but if it's the 2nd\3rd scatter this won't be the same as kinc. Perhaps I should call this kbefore to distinguish it from kinc.

RichardWaiteSTFC replied:
No I think `kinc` is good when talking about the incident k-vector in a scattering event regardless of the order of the scattering, just would be good to distinguish experimental kfixed from ki/kf of a scattering event - but variable names can only do so  much!

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -23,17 +26,48 @@
 constexpr int DEFAULT_LONGITUDINAL_DETS = 10;
+
+/// Energy (meV) to wavevector (angstroms-1)
+inline double toWaveVector(double energy) { return sqrt(energy / PhysicalConstants::E_mev_toNeutronWavenumberSq); }
```
Is there no central place in mantid that stores these unit conversions so we don't have them repeated throughout the code?

DannyHindson replied:
Good question (especially as I'm normally the cheerleader for the Unit class!)
So the reason I've got these DIY functions at the top of the unit is that this algorithm is loosely based on MonteCarloAbsorption.cpp and that has unit conversion functions at the top (it has toWavelength). But that's just passing the buck...
There is some similar code in Unit.cpp here (see line where a quick conversion to "Momentum" is specified):
```
Energy::Energy() : Unit(), factorTo(DBL_MIN), factorFrom(DBL_MIN) {
  addConversion("Energy_inWavenumber", PhysicalConstants::meVtoWavenumber);
  const double toAngstroms = 1e10;
  const double factor =
      toAngstroms * PhysicalConstants::h / sqrt(2.0 * PhysicalConstants::NeutronMass * PhysicalConstants::meV);
  addConversion("Wavelength", factor, -0.5);
  addConversion("Momentum", 2 * M_PI / factor, 0.5);
}
```
In order to use these quick conversions you have to retrieve the factor and power and then write something like this in your own code `factor * std::pow(x, power)`. So a square root gets implemented as pow(x, 0.5) and I believe this is often slower than the sqrt function:
https://stackoverflow.com/questions/11810686/is-fast-implementation-of-powx-0-5f-faster-than-fast-sqrtx

When I've run this Discus algorithm through the performance profiler the sqrt function has shown up as a pinch point. So I think I'd prefer to leave things as they are rather than make the square roots even slower. But to be honest I've only just researched this. Probably worth a comment next to the DIY functions to justify their use at least.

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -23,17 +26,48 @@
+private:
+  const ExperimentInfo &m_expt;
+  const DeltaEMode::Type m_emode;
+  double m_value;
```
what is `m_value` here? looks like something to do with the energy mode?

DannyHindson replied:
This struct\class is also copied from MonteCarloAbsorption.cpp. `m_value` is the EFixed value on a direct instrument. So for an elastic instrument or an indirect (where it seems the EFixed value can potentially vary by detector) it's not used. Naming could be improved by calling it m_EFixed or similar. I'll update it. In an ideal world perhaps this bit of code could also be shared between the two algorithms but since it's only ~10 lines it's perhaps not worth having a separate .h\.cpp file

RichardWaiteSTFC replied:
Thanks, `m_EFixed` sounds good to me - agreed that ~10 lines is not worth a separate file!

SpencerHowells replied:
For Indirect EFixed only depends on the analyser & reflection. It does not vary by detector.

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -146,32 +176,58 @@ std::map<std::string, std::string> DiscusMultipleScatteringCorrection::validateI
+          "SimulateEnergiesIndependently is only applicable to inelastic direct geometry calculations";
+    if (inputWS->getEMode() == Kernel::DeltaEMode::Indirect)
+      issues["SimulateEnergiesIndependently"] =
+          "SimulateEnergiesIndependently is only applicable to inelastic direct geometry calculations. Different "
```
Perhaps the fact the property `SimulateEnergiesIndependently` applies only to direct geometry could be mentioned in the doc string in  `declareProperty`

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -208,6 +264,74 @@ void DiscusMultipleScatteringCorrection::getXMinMax(const Mantid::API::MatrixWor
+    m_SQWS->replaceAxis(1, std::move(newAxis));
+  }
+  // avoid repeated conversion of bin edges to points inside loop by converting to point data
+  convertWsToPoints(m_SQWS);
```
This this is repeated a few lines before (L269)

DannyHindson replied:
It's supposed to be called again if the axes on the workspace have been transposed so that both axes (Q,w) are converted to points. But it's not in the right place. Should be at the end of the first bit of the if...else statement so I'll move it there.

DannyHindson replied:
Just noticed that convertWSToPoints actually converts both axis (x axis and spectrum axis) so no need to call it twice at all. I'll remove the second call

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -208,6 +264,74 @@ void DiscusMultipleScatteringCorrection::getXMinMax(const Mantid::API::MatrixWor
+
+    std::transform(ySQ.begin(), ySQ.end(), ySQ.begin(), [](double d) -> double {
+      if (d == 0.)
+        return -20.0;
```
What's the significance of -20 here, is it just an unphysical value?

DannyHindson replied:
This bit of code is copied from the original Discus Fortran code. It's just a large negative value where exp(-20) is considered to be close to zero. It gets around generating an infinity when you take the log of a workspace containing zeroes. I think it deserves a comment or perhaps constant rather than just an arbitrary looking number in the code.

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -208,6 +264,74 @@ void DiscusMultipleScatteringCorrection::getXMinMax(const Mantid::API::MatrixWor
+  }
+}
+
+void DiscusMultipleScatteringCorrection::convertWsToPoints(MatrixWorkspace_sptr &ws) {
```
IS there a reason not to use the algorithm `ConvertToPointData`?

DannyHindson replied:
When importance sampling is "on" there is some flat interpolation done later on in the main bit of the algorithm. So when converting a histogram S(Q) to points it's better to store the y values at the x value of the left bin edge rather than the x value of the bin centre (which is what ConvertToPointData does). Without this you get an offset in the results where the S(Q) appears to be offset by half a bin - I must have seen this in some calculations with fat bins.

This was a choice I'd made before this PR. I think it's just shown up in the diff here because I decided to prune down the algorithm's exec method because it was getting quite large so this (pre-existing) code is now in its own function

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -710,39 +1037,103 @@ DiscusMultipleScatteringCorrection::scatter(const int nScatters, Kernel::PseudoR
+  double qrange;
+  const double qmin = abs(kf - ki);
+  if (ki <= kf) {
+    qrange = 2 * ki;
```
So presumably |ki| and |kf| are determined from the energy transfer. I think this is getting the |q| range given |ki| and |kf|, the minimum being when ki and kf are parallel (|ki|-|kf|  - is using inelastic convention), the maximum being when they are anti-parallel (|ki| + |kf|)? The range being the difference of the two. But I had a couple of questions:

1.  Why is the abs value of qmin taken (surely it doesn't have to be positive)?
2. How is this effected by the crystallography convention (for example in the convention use above qrange = 2*kf, using the opposite convention it would be 2*ki, here your code seems to determine which is the appropriate one from the relative magnitudes of |ki| and |kf|, why is that the case?

DannyHindson replied:
So yes the code generally handles each scatter by starting with ki, it samples an energy transfer to get kf and then works out q from ki and kf.
In the Discus code q is always >=0 so I've got the same here. I think this is done because in S(q) and S(q,w) for powders and liquids they are really S(|q|) and S(|q|,w) with only the w axis ranging across negative and positive values.
I did have a look around the crystallography convention stuff in Mantid based on a chat we had in one of our standups a couple of weeks ago. It seems like the q sign is mainly used in code that determines hkl values. Perhaps if this multiple scattering correction was ever extended to work on single crystals the assumption that q>=0 might need to change but not attempting to cover that type of instrument so far.
I'll change the code to use std::min at least but let me know if this doesn't sound right.

DannyHindson replied:
One other thing about the crystallography convention\flag in Mantid. While looking into it I didn't see anything which linked this setting to the energy transfer calculation. So a positive energy transfer in Mantid is always an energy loss regardless of the crystallography convention. If there was ever any inelastic single crystal logic in Mantid this might cause a problem - not sure if that exists as a type of experiment though.

RichardWaiteSTFC replied:
> So yes the code generally handles each scatter by starting with ki, it samples an energy transfer to get kf and then works out q from ki and kf. In the Discus code q is always >=0 so I've got the same here. I think this is done because in S(q) and S(q,w) for powders and liquids they are really S(|q|) and S(|q|,w) with only the w axis ranging across negative and positive values. I did have a look around the crystallography convention stuff in Mantid based on a chat we had in one of our standups a couple of weeks ago. It seems like the q sign is mainly used in code that determines hkl values. Perhaps if this multiple scattering correction was ever extended to work on single crystals the assumption that q>=0 might need to change but not attempting to cover that type of instrument so far. I'll change the code to use std::min at least but let me know if this doesn't sound right.

OK sorry the convention in mantid settings is not relevant here,  it's whatever you decide in the code/workspace that you give, but it won't matter because presumably you only supply S(|q|,w) toi sample anyway (and safe to assume S(q,w) = S(-q,w)).

RichardWaiteSTFC reviewed (commented):
Nice work, script works well and provides the correct workspaces. I just have a few questions.

DannyHindson committed:
Incorporate review comments

Clarify k variable names in a couple of places
Explain why using local unit conversion code for k to E
Use std::min rather than explicit if statement
Remove redundant call to convertWSToPoints and clarify that this
function actually converts both axes to points

DannyHindson committed:
Tighten up validation on S(Q,w) for inelastic calculations

If S(Q,w) only contained values for positive w (energy loss) then some calculations were
failing in SampleKW because there were no available w values in S(Q,w) that allowed a
path to be completed if the neutron's energy had already dropped close to zero
Fix by requiring S(Q,w) to contain some negative w (energy gain) values and add unit test
to cover this. Annoyingly one of the calculation result tests used an S(Q,w) with only positive
values so I've also updated that so that the spike in the test S(Q,w) is at -ve rt +ve w

Also spotted that the importance sampling for inelastic assumed the S(Q,w) had constant
bin width in w. This is the case in all the test inputs I've used but might not always
be so have multiplied by the w bin width in the code when calculating the probability
distribution for Q, w

Also tidy up use of braces in ValidateInputs on single line if statements and append to
issues list rather than overwrite in validations on StructureFactorWorkspace

DannyHindson committed:
Resolve compiler warning

DannyHindson committed:
Rename variable now that qmax not always 2k

DannyHindson committed:
Performance improvement to only do sqrt once in interpolation

RichardWaiteSTFC reviewed (dismissed):
Thanks for validation and variable name changes,also good catch on accounting for energy bin-width when doing importance sampling.

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -208,6 +287,77 @@ void DiscusMultipleScatteringCorrection::getXMinMax(const Mantid::API::MatrixWor
+    }
+  }
+  if (dynamic_cast<BinEdgeAxis *>(ws->getAxis(1)) != nullptr) {
+    auto edges = dynamic_cast<NumericAxis *>(ws->getAxis(1))->getValues();
+    std::vector<double> centres(edges.size() - 1);
+    for (size_t i = 0; i < centres.size(); ++i) {
+      centres[i] = (0.5 * (edges[i] + edges[i + 1]));
```
I'm pretty sure there's handling from edges -> centres in Mantid:

Calling `Histogram::Points` should avoid having to check and convert: https://docs.mantidproject.org/nightly/concepts/HistogramData.html#points

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -219,46 +369,53 @@ void DiscusMultipleScatteringCorrection::exec() {
+  }
+  m_QSQWS = prepareQSQ(qmax);
+  size_t totalPoints = 0;
+  for (size_t i = 0; i < m_QSQWS->getNumberHistograms(); i++) {
+    totalPoints += m_QSQWS->histogram(i).size();
+  }
```
Use std::accumulate to appease the CppCheck gods in the future

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -219,46 +369,53 @@ void DiscusMultipleScatteringCorrection::exec() {
+  // call this function with dummy efixed to determine total possible simulation points
+  const auto inputNbins = generateInputKOutputWList(-1.0, inputWS->points(0).rawData()).size();
+
+  int nSimulationPointsInt = getProperty("NumberOfSimulationPoints");
+  size_t nSimulationPoints = static_cast<size_t>(nSimulationPointsInt);
```
This can be done as a single step:
```c++
auto nSimulationPoints = static_cast<size_t>(getProperty("NumberOfSimulationPoints"));
```

DannyHindson replied:
I did try this initially but it gives this error when you supply a value for this parameter:
`Attempt to assign property NumberOfSimulationPoints to incorrect type. Expected type uint64_t`
It seems like the result of a call to getProperty needs to be assigned to an int variable initially because the parameter is set up as an int in the algorithm's Init method.
Having a size_t parameter seems quite hard work in Mantid and is something I've come across before on other algorithms. If you instead set it up as a size_t in init, it makes supplying values a bit of a pain because there aren't size_t literals. So hard to write a C++ unit test for example if the parameter is set up as size_t

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -219,46 +369,53 @@ void DiscusMultipleScatteringCorrection::exec() {
+  int nSimulationPointsInt = getProperty("NumberOfSimulationPoints");
+  size_t nSimulationPoints = static_cast<size_t>(nSimulationPointsInt);
+  if (isEmpty(nSimulationPoints) || nSimulationPoints > inputNbins) {
+    if (!isEmpty(nSimulationPoints)) {
+      g_log.warning() << "The requested number of simulation points is larger "
+                         "than the maximum number of simulations per spectra. "
+                         "Defaulting to "
+                      << inputNbins << ".\n ";
```
If the simulation points are empty we don't emit a warning?
If we don't warn for an empty number of points we can simplify the if.

If we should check a better flow would be:

```c++
if (isEmpty(nSimulationPoints)){
    g_log.warning("Empty")
    nSimulationPoints = inputNbins;
} else if (nSimulationPoints > inputNbins) {
    g_log.warning(" The requested number...")
    nSimulationPoints = inputNbins;  // Intentionally duplicated as simpler flow > DRY principle for 2 statements
}
```

DannyHindson replied:
nSimulationPoints is an optional parameter so fine not to supply it and no warning required in that case

DavidFair replied:
Can we simplify the if from:
```c++
  if (isEmpty(nSimulationPoints) || nSimulationPoints > inputNbins) {
    if (!isEmpty(nSimulationPoints)) {
```
To:
```c++
if (isEmpty(nSimulationPoints)){ 
    nSimulationPoints = inputNbins; 
} else if (nSimulationPoints > inputNbins) { 
    g_log.warning(" The requested number...") 
    nSimulationPoints = inputNbins;
}
```

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -428,110 +599,219 @@ void DiscusMultipleScatteringCorrection::exec() {
+      index++;
+      return t;
+    });
+  } else {
```
Return kInW from this if branch and remove the else to remove a layer of nesting

DannyHindson replied:
As a point of style I prefer having the single point of return in this case because the different branches are genuine alternatives ie the first "if branch" isn't some kind of pre-validation before proceeding to the real work in the later branches. If I chose in future to post-process the result (eg filter it or sort it) I can also do it once at the moment.
If there's some kind of general rule either in Mantid or the wider C++ world that I'm going against though I'm happy to be told otherwise

DavidFair replied:
I mean it's more like emacs vs vim in the C++ world so I'm not going to get too involved on this one haha.

My concern is the nested logic in the else branch. It feels like there are 2-3 private methods rolled into a single method for INDIRECT / DIRECT / ELASTIC that could be unrolled out (the one liner for DIRECT should stay at the moment)

DannyHindson replied:
It feels like those lines belong together. The Direct case is actually split into two different approaches depending on the `SimulateEnergiesIndependently` flag (one in the for loop on L628 and one outside on L624). I think seeing all the different tuples that get put into the kInW vector in one function is also useful for seeing what this vector means\does.

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -615,6 +899,80 @@ double DiscusMultipleScatteringCorrection::interpolateFlat(
+ */
+double DiscusMultipleScatteringCorrection::Interpolate2D(MatrixWorkspace_sptr SOfQ, double w, double q) {
+  double SQ;
+  int iW;
```
Set these values to the safe value up-front so someone doing some overzelous refactoring in the future is protected:
```c++
double SQ = 0;
int iW = -1;
```

...

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -615,6 +899,80 @@ double DiscusMultipleScatteringCorrection::interpolateFlat(
+  try {
+    // required w values will often equal the points in the S(Q,w) distribution so pick nearest value
+    iW = static_cast<int>(Kernel::VectorHelper::indexOfValueFromCentersNoThrow(wValues, w));
+  } catch (std::out_of_range) {
+    iW = -1;
+  }
```
Then this line can be done as
```
catch(std::out_of_range &){}  // Note ref to avoid triggering a copy here
```

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -615,6 +899,80 @@ double DiscusMultipleScatteringCorrection::interpolateFlat(
+      SQ = interpolateFlat(SOfQ->getSpectrum(iW), q);
+    else
+      SQ = interpolateGaussian(m_logSQ->getSpectrum(iW), q);
+  } else
+    SQ = 0.;
```
Ditto this else can also be dropped too that way

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -710,39 +1078,106 @@ DiscusMultipleScatteringCorrection::scatter(const int nScatters, Kernel::PseudoR
+    // slightly concerned that rounding errors moving between k and E may mean we take the sqrt of
+    // a negative number in here. deltaE was capped using a threshold calculated using fromWaveVector so
+    // hopefully any rounding will affect fromWaveVector(kinc) in same direction
+    kf = toWaveVector(fromWaveVector(kinc) - deltaE);
```
Could we put an assert in to protect from this without incurring a runtime penalty for most users?
That way if it every crashes due to imaginary values a developer running Debug mode will really quickly track it down

DavidFair reviewed (changes requested):
Partial review of DiscusMultipleScatteringCorrection `.cpp` and `.h`

DannyHindson commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -45,23 +81,18 @@ DECLARE_ALGORITHM(DiscusMultipleScatteringCorrection)
-  auto wsQValidator = std::make_shared<CompositeValidator>();
-  wsQValidator->add<WorkspaceUnitValidator>("MomentumTransfer");
-
-  declareProperty(std::make_unique<WorkspaceProperty<>>("SofqWorkspace", "", Direction::Input, wsQValidator),
```
Yes - but the release note indicates it's a breaking change + the algorithm currently has this message at the start:

> DiscusMultipleScatteringCorrection is in the beta stage of development. Its name, properties and behaviour may change without warning.

The logic here was that it was important to get the initial version of the algorithm out there in the nightly build for the diffraction users a few months ago. It was always anticipated that feedback from those users or the addition of inelastic support might involve some breaking changes

DavidFair replied:
Ah I didn't see that warning, 👍

DannyHindson commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -138,38 +170,85 @@ std::map<std::string, std::string> DiscusMultipleScatteringCorrection::validateI
+
+  const bool simulateEnergiesIndependently = getProperty("SimulateEnergiesIndependently");
+  if (simulateEnergiesIndependently) {
+    if (inputWS->getEMode() == Kernel::DeltaEMode::Elastic)
+      issues["SimulateEnergiesIndependently"] =
+          "SimulateEnergiesIndependently is only applicable to inelastic direct geometry calculations";
+    if (inputWS->getEMode() == Kernel::DeltaEMode::Indirect)
+      issues["SimulateEnergiesIndependently"] =
+          "SimulateEnergiesIndependently is only applicable to inelastic direct geometry calculations. Different "
```
I'd separated it out because the message is slightly different for Elastic and Indirect modes. The Indirect message includes an extra bit of explanation that's not relevant to someone running on an Elastic workspace

DavidFair commented on docs/source/algorithms/DiscusMultipleScatteringCorrection-v1.rst:
```diff
@@ -130,18 +143,222 @@ The sample shape can be specified by running the algorithms :ref:`SetSample <alg
+   Sofq_isotropic=CreateWorkspace(DataX=X,DataY=Y,UnitX="MomentumTransfer")
+
+   two_thetas=[]
+   for i in range(180):
```
Should this be 0-179 or 0-180? I think there might be an off-by-one error in this example?

DannyHindson replied:
Yes having the range as 0-180 might have been nicer but I think it needs to be 0-179 because in the call to CreateSampleWorkspace I rely on the fact that the number of detectors is divisible by 4. So I need 180 detectors rather than 181.

CreateSampleWorkspace has a validation on the NumBanks parameter that means it has to be <=100 so I couldn't just supply 180 (or 181) in this parameter. So I divide by four and make each bank contain 4 detectors.

DannyHindson committed:
Fix problem in validation of S(Q,w)

Wrong index was used to retrieve w

DannyHindson commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -138,38 +170,85 @@ std::map<std::string, std::string> DiscusMultipleScatteringCorrection::validateI
+            wValue = SQWS->getAxis(1)->getValue(iBin);
+          }
+          if ((SQWS->y(iHist)[iBin] > 0.) && (wValue < 0.))
+            atLeastOnePositive = true;
```
Just noticed there's actually a bug in here as well. L216 should be using the other index ie
`SQWS->getAxis(1)->getValue(iHist)`
Re using `std::any_of`, this check steps through two vectors of w values and y values so will need to create a vector of pairs to pass into `std::any_of`. The resulting code might not be that concise but will at least avoid the CppCheck warning

DavidFair replied:
If you think it impacts readability significantly feel free to leave this one out and we can simply suppress it

DannyHindson committed:
Incorporate review comments

Rearrange validation on S(Q,w) to rely on std::any_of
Rely on existing Mantid code to calculate bin centres from edges
Calculate total points in QSQ workspace by calling size() method
(which internally uses std::accumulate)
Add assert to check kf never imaginary (nan)

DannyHindson committed:
Make nSimulationPoints logic more readable

DavidFair reviewed (approved):
I've retested this and it still works
Thanks for doing those changes, lets :shipit:
